src/allmydata/watchdog/inotify.py

```diff
@@ -0,0 +1,139 @@
+
+from watchdog.observers import Observer
+from watchdog.events import FileSystemEventHandler, FileSystemMovedEvent, FileModifiedEvent, DirModifiedEvent, FileCreatedEvent, FileDeletedEvent, DirCreatedEvent, DirDeletedEvent
+
+from twisted.internet import reactor
+from twisted.python.filepath import FilePath
+from twisted.python.filepath import InsecurePath
+
+from allmydata.util.pollmixin import PollMixin
+from allmydata.util.assertutil import _assert, precondition
+from allmydata.util import log, encodingutil
+from allmydata.util.encodingutil import unicode_from_filepath
+from allmydata.util.fake_inotify import humanReadableMask, \
+    IN_WATCH_MASK, IN_ACCESS, IN_MODIFY, IN_ATTRIB, IN_CLOSE_NOWRITE, IN_CLOSE_WRITE, \
+    IN_OPEN, IN_MOVED_FROM, IN_MOVED_TO, IN_CREATE, IN_DELETE, IN_DELETE_SELF, \
+    IN_MOVE_SELF, IN_UNMOUNT, IN_Q_OVERFLOW, IN_IGNORED, IN_ONLYDIR, IN_DONT_FOLLOW, \
+    IN_MASK_ADD, IN_ISDIR, IN_ONESHOT, IN_CLOSE, IN_MOVED, IN_CHANGED
+[humanReadableMask, \
+    IN_WATCH_MASK, IN_ACCESS, IN_MODIFY, IN_ATTRIB, IN_CLOSE_NOWRITE, IN_CLOSE_WRITE, \
+    IN_OPEN, IN_MOVED_FROM, IN_MOVED_TO, IN_CREATE, IN_DELETE, IN_DELETE_SELF, \
+    IN_MOVE_SELF, IN_UNMOUNT, IN_Q_OVERFLOW, IN_IGNORED, IN_ONLYDIR, IN_DONT_FOLLOW, \
+    IN_MASK_ADD, IN_ISDIR, IN_ONESHOT, IN_CLOSE, IN_MOVED, IN_CHANGED]
+
+
+TRUE  = 0
+FALSE = 1
+
+NOT_STARTED = "NOT_STARTED"
+STARTED     = "STARTED"
+STOPPING    = "STOPPING"
+STOPPED     = "STOPPED"
+
+class INotifyEventHandler(FileSystemEventHandler):
+
+    def __init__(self, path, mask, callbacks, pending_delay):
+        print "init INotifyEventHandler"
+        FileSystemEventHandler.__init__(self)
+        self._path = path
+        self._mask = mask
+        self._callbacks = callbacks
+        self._pending_delay = pending_delay
+        self._pending = set()
+
+    def process(self, event):
+        event_filepath_u = event.src_path.decode(encodingutil.get_filesystem_encoding())
+
+        if event_filepath_u == unicode_from_filepath(self._path):
```

Comment:
This is probably too specific because there's no guarantee that the paths are canonicalized.